File: api/schema.py
import graphene
from django.contrib.auth import get_user_model
from graphene import ObjectType
from graphene_django import DjangoListField
from graphene_django import DjangoObjectType
from graphql_auth import mutations
from graphql_auth.schema import UserQuery, MeQuery

from blog.models import Category, Article, Like


class AuthMutation(graphene.ObjectType):
    register = mutations.Register.Field()
    verify_account = mutations.VerifyAccount.Field()
    resend_activation_email = mutations.ResendActivationEmail.Field()
    send_password_reset_email = mutations.SendPasswordResetEmail.Field()
    password_reset = mutations.PasswordReset.Field()
    password_set = mutations.PasswordSet.Field()
    password_change = mutations.PasswordChange.Field()
    update_account = mutations.UpdateAccount.Field()
    archive_account = mutations.ArchiveAccount.Field()
    delete_account = mutations.DeleteAccount.Field()
    send_secondary_email_activation = mutations.SendSecondaryEmailActivation.Field()
    verify_secondary_email = mutations.VerifySecondaryEmail.Field()
    swap_emails = mutations.SwapEmails.Field()
    remove_secondary_email = mutations.RemoveSecondaryEmail.Field()

    token_auth = mutations.ObtainJSONWebToken.Field()
    verify_token = mutations.VerifyToken.Field()
    refresh_token = mutations.RefreshToken.Field()
    revoke_token = mutations.RevokeToken.Field()


class CategoryType(DjangoObjectType):
    class Meta:
        model = Category
        fields = ('id', 'name', 'slug')


class LikeType(DjangoObjectType):
    class Meta:
        model = Like
        fields = ('id', 'user', 'article', 'created_on')


class UserType(DjangoObjectType):
    class Meta:
        model = get_user_model()
        fields = (
            'id', 'username', 'first_name', 'last_name', 'email', 'is_active', 'is_staff', 'is_superuser',
            'groups')


class ArticleType(DjangoObjectType):
    user = UserType
    likes = DjangoListField(LikeType)

    class Meta:
        model = Article
        fields = ('id', 'slug', 'title', 'content', 'user', 'likes', 'categories', 'created_on', 'updated_on')


class Query(UserQuery, MeQuery, ObjectType):
    categories = DjangoListField(CategoryType)
    articles = DjangoListField(ArticleType)
    users = DjangoListField(UserType)


class Mutation(AuthMutation, ObjectType):
    ...


schema = graphene.Schema(query=Query, mutation=Mutation)

File: frontend/src/schema.ts
import gql from "graphql-tag";

export const typeDefs = gql`
type User {
    id: ID!
    username: String!
    firstName: String
    lastName: String
    likes: [Like]
}

type Category {
    id: ID!
    slug: String!
    name: String!
}

type Like {
    user: User!
    article: Article!
    createdOn: String!
    
}

type Article {
    id: ID!
    slug: String!
    title: String!
    content: String!
    user: User!
    categories: [Category]
    likes: [Like]
    createdOn: String!
    updatedOn: String!
}

type Error {
    message: String
    locations: [ErrorLocations]
    path: [String]
}

type ErrorLocations {
    line: Int
    column: Int
}

type TokenAuthResponse {
    success: Boolean
    token: String
    refreshToken: String
    refreshExpiresIn: Int
    errors: [String]
}

type Mutation {
    tokenAuth(username: String!, password: String!): TokenAuthResponse
}

type Query {
    articles: [Article]
}
`
